### PR TITLE
Fix gh-1400: Button Behavior for Search Result State

### LINKED ISSUE #1400: Load More Button Behavior in Search Page
Some search scenarios might be better have the `Load More` button hidden.

**1. No results were found**

screen shot 2017-07-06 at 6 45 33 pm

Would it be better if there's a copy that says no match was found (or something similar) instead of still showing 'load more'?

**2. The search results are less than 16 (limit per search)**

screen shot 2017-07-06 at 7 33 40 pm

**3. Finally, when there are no more Projects/Studios to load upon clicking the button**

(tested using Chrome and Safari on Mac OSX)

### PR DESCRIPTION
**Pull request by carlflor**
Resolves #1400 

- Shows "prompt message" if search term is blank
- Hides Load More Button when:
  - the search results are less than 16
  - there are no more to be loaded
  - Empty Search result
- Added phrase  "Nothing found" for `general.searchEmpty`

*Test Cases*
`scratch` - Returns search result with `Load More` button
`saturated` - Returns more than 16 matches, but hides `Load More` after a few more requests
`flabbergasted` - Returns 16 matches
`pull request` - Returns less than 16 matches (4)
`clean coder` - No search results

### CODE REVIEW

**Comment by thisandagain** [comment]
Hi @carlflor thanks for the PR! 

It looks like your PR is not passing some tests however. You can run the tests locally yourself by running `npm test` or look at the output from TravisCI. Please let us know once all tests are passing and then we can take another look!

**Comment by carlflor** [comment]
@thisandagain fixed it! sorry about that! :)

**Comment by mewtaylor** [comment]
@carlflor so sorry for the delay in getting to this, was out of town for a bit and playing catch up after – I'll take a look at this later today!

File: src/views/search/search.jsx [hunk 1]
```diff
@@ -70,9 +70,8 @@ var Search = injectIntl(React.createClass({
         }, function (err, body) {
             var loadedSoFar = this.state.loaded;
             Array.prototype.push.apply(loadedSoFar, body);
-            this.setState({loaded: loadedSoFar});
             var currentOffset = this.state.offset + this.props.loadNumber;
-            this.setState({offset: currentOffset});
+            this.setState({loaded: loadedSoFar, offset: currentOffset});
```
Comment (mewtaylor) [inline src/views/search/search.jsx]: nice, thanks for this :)

File: src/views/search/search.jsx [hunk 2]
```diff
@@ -96,6 +95,31 @@ var Search = injectIntl(React.createClass({
         }
         return allTab;
     },
+    getProjectBox: function () {
+        var results = <Grid items={this.state.loaded}
```
Comment (mewtaylor) [inline src/views/search/search.jsx]: We're starting to move to a different style for this kind of multiline – would you mind changing this slightly to something like:
```jsx
var results = <Grid
    items={this.state.loaded}
    itemType={this.props.tab}
    cards={true}
    showAvatar={true}
    showLoves={false}
    showFavorites={false}
    showViews={false}
/>;
```

File: src/views/search/search.jsx [hunk 3]
```diff
@@ -96,6 +95,31 @@ var Search = injectIntl(React.createClass({
...
+                            itemType={this.props.tab}
+                            cards={true}
+                            showAvatar={true}
+                            showLoves={false}
+                            showFavorites={false}
+                            showViews={false} />;
+        var searchAction;
+        if (this.props.searchTerm === '') {
```
Comment (mewtaylor) [inline src/views/search/search.jsx]: We actually, if you look on production, will show results for a general query when the search term is empty. It appears that there is a separate bug right now for that where you need to click `Load More` to display the first set of results though. I'd recommend removing this check, and I can file a separate bug for the empty search issue.

File: src/views/search/search.jsx [hunk 4]
```diff
@@ -96,6 +95,31 @@ var Search = injectIntl(React.createClass({
...
     },
+    getProjectBox: function () {
+        var results = <Grid items={this.state.loaded}
+                            itemType={this.props.tab}
+                            cards={true}
+                            showAvatar={true}
+                            showLoves={false}
+                            showFavorites={false}
+                            showViews={false} />;
+        var searchAction;
```
Comment (mewtaylor) [inline src/views/search/search.jsx]: Just to be explicit would you mind setting this to null, i.e. `var searchAction = null;`?

**Comment by mewtaylor** [comment]
One other thing I forgot to mention – we add strings to be translated to a local `l10n.json` file for a page usually, though as you pointed out @carlflor there is no `l10n.json` file for `search`. If you'd like to add that string to the `general` file, you can add it to the `src/l10n.json` top-level translation string file.

**Comment by mewtaylor** [comment]
@carlflor do you have time to take a look at these changes? If not, no worries – I can close this and start a new PR building on your work with these changes.

**Comment by carlflor** [comment]
@mewtaylor Sorry, it took me a long while to accomplish the changes. Let me know if I need to add more. :)

**Comment by mewtaylor** [comment]
@carlflor no worries at all :) – I may not get to this until Monday, but I'll try to get back to you as soon as I can!

**Comment by mewtaylor** [comment]
@carlflor this looks great! I unfortunately just realized that we have one issue though, which is this: https://github.com/LLK/scratch-www/issues/575#issuecomment-226187539

This bug is still around, we are just tracking it in another repository. I wonder if, as a consequence of that bug still being around, we'd need to change this to look only at the length of `body` returned by the request, and see if that length is less than the offset. In the bug mentioned above, the response from the api will still return 16 projects if there are more, but it will overlap with some existing projects and in that overlap will be where the number. 

I am trying to reproduce the bug right now, and having difficulty. If I can't reproduce it in the next week or so, then we could probably merge this in as-is. Thanks for your patience on this!

**Comment by carlflor** [comment]
Sure! No worries. :)

**Comment by Kenny2github** [comment]
@mewtaylor this is waaaay past milestone - any updates?

**Comment by mewtaylor** [comment]
@carlflor sorry for the delay on this – we were having trouble reproducing the bug I had mentioned a few months back, but have recently reproduced it.

Given that, I think the best direction forward with this behavior would be to do this check in the api response, i.e. to see if the length of the body returned by the api request is less than `this.props.loadNumber`. That should help alleviate the issue by isolating the check to the api call, rather than the concatenated list.

I know it's been a while since I've been able to get back to you on this, so if you do not have the capacity to look into this now, no worries! I'd be happy to take it from here if that would be helpful for you, but if you're up to make this adjustment, that'd be awesome (and I'm around to answer any questions you might have too).

**Comment by carlflor** [comment]
Hi @mewtaylor , I'll take a look within the week. Would that be fine?

**Comment by mewtaylor** [comment]
@carlflor absolutely! Thanks!

**Comment by carlflor** [comment]
Hi @mewtaylor , apologies for the delay. Could you check if the changes work for the issue?

File: src/views/search/search.jsx [hunk 5]
```diff
@@ -70,9 +71,9 @@ var Search = injectIntl(React.createClass({
         }, function (err, body) {
             var loadedSoFar = this.state.loaded;
             Array.prototype.push.apply(loadedSoFar, body);
-            this.setState({loaded: loadedSoFar});
             var currentOffset = this.state.offset + this.props.loadNumber;
-            this.setState({offset: currentOffset});
+            var willLoadMore = body.length == this.props.loadNumber;
```
Comment (mewtaylor) [inline src/views/search/search.jsx]: this should be `===` rather than `==`

**Comment by carlflor** [comment]
@mewtaylor done! :)

File: src/l10n.json [hunk 6]
```diff
@@ -62,6 +62,7 @@
     "general.scratchJr": "ScratchJr",
     "general.scratchStore": "Scratch Store",
     "general.search": "Search",
+    "general.searchEmpty": "Nothing found",
```
Comment (mewtaylor) [inline src/l10n.json]: @thisandagain @christanbalch @ntlrsk thoughts on this language for an empty search page?

**Comment by carlflor** [comment]
Cool! Thanks :)

File: src/l10n.json (hunk 6 above)
Comment (mewtaylor) [reply src/l10n.json]: @thisandagain @christanbalch @ntlrsk just wanted to check one more time about empty search language – does this look alright to you?

**Comment by redmangospros** [comment]
is this running, @mewtaylor

File: src/l10n.json (hunk 6 above)
Comment (thisandagain) [reply src/l10n.json]: This looks fine for now. Thanks @mewtaylor

**Comment by redmangospros** [comment]
Great :)

**Comment by Kenny2github** [comment]
@mewtaylor getting close to milestone, any updates?

**Comment by rschamp** [comment]
After merging the ES6 upgrade (#1790), there will be conflicts in the outstanding PRs. Because it will be complicated to merge, we will fix the conflicts ourselves, so don't worry about trying to fix it unless you want to.

**Comment by carlflor** [comment]
Feel free to modify it, I might not be able to look into it any time soon. :)

**Comment by rschamp** [comment]
Thank you @carlflor!

**Pull request merged**